Add Crust to Allwinner boards to enable power management functions

viraniac opened the pull request:
# Description

Add support to build Crust for Allwinner boards to enable power management functions

Jira reference number [AR-1690]

# How Has This Been Tested?

- [X] Booted on Nanopiduo2. U-boot reports starting of crust and also shows crust version

Tests performed using crust's debug monitor. Enabled using following patch placed in userpatches/crust
```
diff --git a/configs/nanopi_m1_plus_defconfig b/configs/nanopi_m1_plus_defconfig
index 77aaea5..d002d61 100644
--- a/configs/nanopi_m1_plus_defconfig
+++ b/configs/nanopi_m1_plus_defconfig
@@ -1,2 +1,4 @@
 CONFIG_PLATFORM_H3=y
 CONFIG_REGULATOR_GPIO_DRAM=y
+CONFIG_DEBUG_MONITOR=y
+CONFIG_DEBUG_PRINT_SPRS=y
```
- [X] Ran `systemctl suspend`. Waited for the system to go to suspended state. Pressed w on the crust debug console and it brought back linux from suspended state.
- [X] Ran `poweroff`. Waited for the system to go to poweroff state. Pressed w on the crust debug console and it started the boot process.

Tests performed with nanopi duo2 connected to power through a [usb multimeter](https://robu.in/product/usb-current-voltage-ammeter-tester/)
- [X] Put board to sleep for 30 seconds using `rtcwake -m mem -s 30` and measured the current. 
  - Current draw in official armbian image without Crust `0.09 Amps`
  - Current draw in armbian image with Crust `0.00 Amps`
- [X] Powered off the board using `poweroff` command and measured the current
  - Current draw in official armbian image without Crust `0.13 Amps`
  - Current draw in armbian image with Crust `0.00 Amps`

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules


[AR-1690]: https://armbian.atlassian.net/browse/AR-1690?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

viraniac commented:
Changes compared to previous version
- Fixed patch creation for crust
- Moved crust related config from config/source/armhf.conf and config/source/arm64.conf to config/source/families/include/sunxi_common.inc and  config/source/families/include/sunxi64_common.inc respectively

The-going commented:
> Powered off the board using `poweroff` command and measured the current
> 
>     * Current draw in official armbian image without Crust `0.13 Amps`
> 
>     * Current draw in armbian image with Crust `0.00 Amps`

Fantastic! You checked it in another way available.

I think I can check and add for A83T, A64 processors.

viraniac commented:
Changes that are made today

- Rebased on latest armbian build
- Moved Crust related U-boot patches to u-boot-sunxi-crust directory. Now they only apply if crust_config is mentioned in board config
- Also replaced my enable_remoteproc_for_h3.patch where I was enabling remoteproc manually for each h3 device with Jernej's version of patch from [LibreELEC's git tree](https://github.com/LibreELEC/LibreELEC.tv/blob/master/projects/Allwinner/patches/u-boot/0012-DO-NOT-MERGE-sunxi-Enable-remoteproc-on-H3.patch)

viraniac commented:
With the changes of this merge request, enabling crust should be as simple as mentioning crust_config in the config/boards/<board_config> file.

@igorpecovnik Would you prefer me to enable crust for all devices that are based on a23, a64, a83t, h3 or h6.  Or should we let individual board maintainers to test and enable the crust individually for their boards?

igorpecovnik commented:
> With the changes of this merge request, enabling crust should be as simple as mentioning CRUSTCONFIG

We can merge this at once, just duo2 will be enabled / covered? That is fine.

> Would you prefer me to enable crust for all devices that are based on a23, a64, a83t, h3 or h6? 

Yes, lets do that, but one by one, to make sure it at least build. I would propose to bump u-boot to v23.07 (now its possible) while doing this and remove as much patches as possible (at this particular moment: BOOTBRANCH="tag:v2023.07-rc4", release within two weeks)

> Or should we let individual board maintainers to test and enable the crust individually for their boards?

Lets call maintainers and interestent to be aware of this commit.
@julialongtin @Tonymac32 @AGM1968 @schwar3kat @engineer-80 @PanderMusubi @sgjava @lbmendes @teknoid @AaronNGray @Kreyren @devdotnetorg

igorpecovnik commented on lib/library-functions.sh:
```diff
@@ -307,6 +307,15 @@ set -o errexit  ## set -e : exit the script if any statement returns a non-true
 # shellcheck source=lib/functions/compilation/atf.sh
 source "${SRC}"/lib/functions/compilation/atf.sh
 
+# no errors tolerated. invoked before each sourced file to make sure.
```
# This file is/was autogenerated by lib/tools/gen-library.sh; don't modify manually

Was this generated or edited?

viraniac replied:
Nice catch. That was edited. Now its generated

viraniac commented:
> I would propose to bump u-boot to v23.07 (now its possible)

I will start working on this. Will bump it one version at a time, just to make sure to catch any patch application or compilation failures

viraniac commented:
> We can merge this at once, just duo2 will be enabled / covered? That is fine.

yes, right now only duo2 will have crust support as only its board config has CRUSTCONFIG specified

igorpecovnik commented:
> I will start working on this.

Withing this PR or we merge this and you will make a new one?

> yes, right now only duo2 will have crust support as only its board config has CRUSTCONFIG specified

OK

viraniac commented:
> Withing this PR or we merge this and you will make a new one?

I was thinking do it as part of a new PR. I see no reason to block this PR to wait for new U-boot

schwar3kat commented:
> With the changes of this merge request, enabling crust should be as simple as mentioning CRUSTCONFIG in the config/boards/<board_config> file.

I'm seeing board nanopiduo2.csc with:
BOOTCONFIG="nanopiduo2_defconfig"
CRUSTCONFIG="nanopi_m1_plus_defconfig"
This is confusing. Is this correct?

orangepi-r1.csc does not build with:
CRUSTCONFIG="orangepi-r1_defconfig"
 
What would be required in the board file (and other changes) to correctly test this on orangepi-r1 (H5) if it can be tested.

viraniac commented:
@schwar3kat The CRUSTCONFIG is name of defconfig file to use that is present in [configs directory](https://github.com/crust-firmware/crust/tree/master/configs) of crust firmware.

For nanopi-r1, as its based on H3, you can try specifying same filename as I have specified for nanopiduo2. If you want to use a custom defconfig, you can add a patch to create the same in patch/crust directory

viraniac commented:
Just noticed, I forgot to initialize CRUSTPATCHDIR variable. I will raise a pull request with the fix.

viraniac commented:
> Just noticed, I forgot to initialize CRUSTPATCHDIR variable. I will raise a pull request with the fix.

Second thought, I will just leave it uninitialized for now as that makes the patch directory as patch/crust. If I will initialize it, it will be something like patch/crust/crust or something which I think is not required in case of crust as its only applicable for allwinner boards. If required, we can initialize the variable in the future

schwar3kat commented:
> Not sure why I read it is nanopi-r1, probably I am still too sleepy.

I must also have been sleepy.  H5 was a typo it should be H2.  
Thanks, I will look at options now that I understand how it works.

viraniac commented:
> > Not sure why I read it is nanopi-r1, probably I am still too sleepy.
> 
> @viraniac - I must also have been sleepy. H5 was a typo it should be H2+. This soc is not listed in the available defconfig's, but I think it is very similar to H3, so I guess it might work with H3 defconfig. I will give it a try. Thanks, I will look at options now that I understand how it works.

For H2+ you can use the same config as H3, try using nanopi_m1_plus_defconfig.

schwar3kat commented:
Well, it doesn't break, so that's good.  But I'm not too sure how to test it.
I wasn't able to get the crust debug console to work like I guessed that it might work after I couldn't hunt out any documentation.   After systemctl suspend the serial uart was dead and I couldn't send w.  I'm guessing that it might need something else.

I used the following patch to successfully create a debug defconfig in the crust-sunxi-mainline cache with the others:

diff --git a/configs/orangepi_r1_defconfig b/configs/orangepi_r1_defconfig
new file mode 100644
index 77aaea5..d002d61 100644
--- /dev/null
+++ b/configs/orangepi_r1_defconfig
@@ -0,0 +1,4 @@
+CONFIG_PLATFORM_H3=y
+CONFIG_REGULATOR_GPIO_DRAM=y
+CONFIG_DEBUG_MONITOR=y
+CONFIG_DEBUG_PRINT_SPRS=y

viraniac commented:
> It didn't wake after suspend... I will investigate some more tomorrow.

Make sure you are waiting until the system is fully suspended. If your board has a power led, wait till it goes off. For me it was taking about 2 seconds and then I was pressing enter to go to crust console.

Best way I found to test its working is to plug a usb multimeter and measure the current draw for the board.

schwar3kat commented:
**OrangePi R1 (H2+) without crust:**
`Memory idling with HTOP: 71.1M + 3.75M`
`systemctl suspend 0.15 Amps`
`poweroff          0.19 Amps`

**OrangePi R1 (H2+) with crust:**
`Memory idling with HTOP: 71.2M + 0.0K (3.65M less!)`
`systemctl suspend 0.15 Amps (no difference. Does not recover with CRUST console w)`
`poweroff          0.10 Amps (0.09 less.  Does recover with CRUST console w)`

3.65M less memory use was unexpected bonus.  
Suspend doesn't appear to use CRUST.  
Poweroff saves 90mA with CRUST.

schwar3kat commented:
If I try to enable CIR and debug, the build fails (firmware too big):
  /usr/lib/gcc/or1k-elf/11/ld: Firmware overflows allocated memory area
  
If I comment out the debug then it builds.
  
diff --git a/configs/orangepi_r1_defconfig b/configs/orangepi_r1_defconfig
new file mode 100644
index 77aaea5..d002d61 100644
--- /dev/null
+++ b/configs/orangepi_r1_defconfig
@@ -0,0 +1,6 @@
+CONFIG_PLATFORM_H3=y
+CONFIG_REGULATOR_GPIO_DRAM=y
+CONFIG_CIR=y
+CONFIG_CIR_USE_OSC24M=y
+CONFIG_DEBUG_MONITOR=y
+CONFIG_DEBUG_PRINT_SPRS=y

viraniac commented:
> If I try to enable CIR and debug, the build fails (firmware too big): /usr/lib/gcc/or1k-elf/11/ld: Firmware overflows allocated memory area
> 
> If I comment out the debug then it builds.

Yeah, probably its size would have become too big to fit into the sram

> **OrangePi R1 (H2+) without CRUST:** `Memory idling with HTOP: 71.1M + 3.75M` `systemctl suspend 0.15 Amps` `poweroff 0.19 Amps`
> 
> **OrangePi R1 (H2+) with CRUST:** `Memory idling with HTOP: 71.2M + 0.0K (3.65M less!)` `systemctl suspend 0.15 Amps (no difference. Does not recover with CRUST console w)` `poweroff 0.10 Amps (0.09 less. Does recover with CRUST console w)`
> 
> 3.65M less memory use was unexpected bonus. Suspend doesn't appear to use CRUST. Poweroff saves 90mA with CRUST.

Interesting. I will try to figure out what may have caused crust not to work in suspend mode but can't make any promises as I don't have the board. Probably it can be some difference between h2+ and h3 chipset. 

How does it behave though with rtcwake? Does running `rtcwake -m mem -s 30` reduces power draw or does that also make no difference?

schwar3kat commented:
> How does it behave though with rtcwake? Does running `rtcwake -m mem -s 30` reduces power draw or does that also make no difference?

Rtcwake doesn't work on this OrangePi R1, but probably because it doesn't  have a hardware rtc and uses software emulation.

I tried to test this on OrangePi zero plus (H5 also no rtc) but unfortunately power management doesn't seem to work on this boards Armbian build.  Systemctl suspend does nothing on a build without crust, and it remains fully powered up. 
Built with crust and the available crust orangepi_zero_plus_defconfig with or without debug, the board doesn't boot.

My other boards are not Allwinner.

viraniac commented:
> Built with crust and the available crust orangepi_zero_plus_defconfig with or without debug, the board doesn't boot.

Could you please try replacing the patches in patch/u-boot/u-boot-sunxi-crust directory with the patches present in the [LibreELEC tree](https://github.com/LibreELEC/LibreELEC.tv/tree/master/projects/Allwinner/patches/u-boot) and see if that works on H5. If it does then I will split the h3-scp patch similar to how it is in the LibreELEC tree.

If it doesn't work, then I am also supposed to get some H5 boards soon. I will be able to debug once they arrive.

viraniac commented:
@schwar3kat Could you please try the image from [here](https://drive.google.com/drive/folders/1JLVDCqr-yxDiyFIkUmyJeGs_yZF4Jw0t?usp=sharing) on your OrangePi zero plus (H5)

schwar3kat commented:
> > 
> 
> Could you please try replacing the patches in patch/u-boot/u-boot-sunxi-crust directory with the patches present in the [LibreELEC tree](https://github.com/LibreELEC/LibreELEC.tv/tree/master/projects/Allwinner/patches/u-boot) and see if that works on H5

I re-did the crust without debug test, and it booted.  I started clean.  I think that there might have been some kind of caching going on with the first build.

Using the patches from LibreELEC behaves identically.  Crust without debug boots, crust with debug does not boot.

`NO CRUST`
`Mem: 95.9M`
`systemctl suspend -doesn't work`
`poweroff: 0.14`

`CRUST`
`Mem: 96.8M`
`systemctl suspend -doesn't work`
`poweroff: 0.16 (slightly worse!)`

`CRUST-LibreELEC`
`Mem: 96.3M`
`systemctl suspend -doesn't work`
`poweroff: 0.16 (slightly worse!)`

I have not tried your image yet. I will do it soon.

I don't think that OrangePi zero plus Armbian image is a good one to test on because power management does not appear to work at all.

schwar3kat commented:
> Could you please try the image from [here](https://drive.google.com/drive/folders/1JLVDCqr-yxDiyFIkUmyJeGs_yZF4Jw0t?usp=sharing) on your OrangePi zero plus

This behaves similar to the other crust's 
DOWNLOAD
Mem: 97.1M
systemctl suspend -doesn't work
poweroff: 0.16
